#106 module date overlapping validation

PhilippeLeopoldie opened the pull request:
This PR improves code clarity, error handling, and user experience through localization and consistent conventions.

- Result class: Moved from LMS.Blazor.Client.Pages.Modules to LMS.Blazor.Client.Models for clearer organization.

- EditModule & CreateModule forms:

   - Refactored form submission to use a new OnSubmit method.
   - HandleValidSubmit now returns Result<bool> for better error handling.
   - Updated OnSubmit to navigate only on successful submission.
   - Improved JSON exception parsing for detailed error messages.
   - Cleaned up commented code and renamed fields for consistency (_moduleState, _navigationManager).

- Exception messages:
   - Translated all user-facing exception messages from English to Swedish for better localization, including overlapping, duplicate, and not-found exceptions.

PhilippeLeopoldie committed:
Refactor Result class namespace

Changed the namespace of the `Result` class from
`LMS.Blazor.Client.Pages.Modules` to
`LMS.Blazor.Client.Models` for better code organization
and clarity regarding its purpose within the project.

PhilippeLeopoldie committed:
Refactor EditModule.razor for improved form handling

- Added using directive for System.Text.Json.
- Changed ModuleForm's OnValidSubmit to use new OnSubmit method.
- Removed commented-out code for fetching module by course ID.
- Modified HandleValidSubmit to return Result<bool> for better error handling.
- Enhanced error handling to parse JSON exception messages.
- Introduced OnSubmit method to manage submission logic and navigation.

PhilippeLeopoldie committed:
Improve form submission and error handling in CreateModule

Refactor `HandleValidSubmit` to return `Result<bool>` for enhanced error reporting. Implement detailed error parsing using `JsonDocument`. Introduce `OnSubmit` method to manage submission flow and navigation based on success. Update `ModuleForm` event handler to utilize the new `OnSubmit` method.

PhilippeLeopoldie committed:
Update error message in InvalidEntryBadRequestException

Changed the error message in the constructor from "Invalid Id: '{id}'" to "Ogiltigt ID: '{id}'" for localization. The parameterless constructor remains unchanged with the message "No patchDocument".

PhilippeLeopoldie committed:
Translate exception messages to Swedish

Updated error messages in the following exception classes:
- ModuleOverlappingException
- ActivityOverlapException
- DuplicateStudentInCourseException
- DuplicateTeacherInCourseException

All messages have been changed from English to Swedish to better serve users.

PhilippeLeopoldie committed:
Translate exception messages to Swedish

Updated the `NotFoundException.cs` file to translate
exception messages for `ModuleNotFoundException`,
`CourseNotFoundException`, and `UserNotFoundException`
from English to Swedish, ensuring localized error handling.

PhilippeLeopoldie committed:
Translate exception messages to Swedish

Updated the exception messages in `UserIsNotTeacherException` and `UserIsNotStudentException` classes from English to Swedish for better localization. The messages now reflect the appropriate language for users.

PhilippeLeopoldie committed:
Refactor service injection and error handling in CreateModule

Reorganized the injection of `IApiService` and `NavigationManager` in `CreateModule.razor`. The `CourseId` parameter remains unchanged. Adjusted error message handling while maintaining the existing form submission logic. Updated the `OnSubmit` method to ensure navigation occurs only on successful submissions.

PhilippeLeopoldie committed:
Refactor EditModule.razor for clarity and consistency

- Reformatted `[Parameter]` attributes for conciseness.
- Renamed `ModuleState` to `_moduleState` for consistency.
- Updated `NavigationManager` to `_navigationManager` and removed duplication.
- Changed `Module` assignment in `OnInitializedAsync` to use `_moduleState`.
- Adjusted formatting in `OnSubmit` method for clarity.

PhilippeLeopoldie committed:
Enhance error message styling in CreateModule.razor

Added top margin to the error message alert for better visual separation from surrounding elements.

PhilippeLeopoldie committed:
Enhance error message styling in EditModule.razor

Added top margin to the error message alert for better visual separation from surrounding elements.

JorgenHagglund commented on Domain.Models/Exceptions/ConflictException.cs:
```diff
@@ -11,31 +11,31 @@ public ConflictException(string message, string title = "Conflict") : base(messa
 public class ModuleOverlappingException : ConflictException
 {
-    public ModuleOverlappingException(string moduleRange) : base($"The time range: {moduleRange} overlaps with an other module in this course.")
+    public ModuleOverlappingException(string moduleRange) : base($"Tidsintervallet: {moduleRange} överlappar med ett annat module i denna kurs.")
```
en annan modul, heter det.

JorgenHagglund commented on Domain.Models/Exceptions/ConflictException.cs:
```diff
@@ -11,31 +11,31 @@ public ConflictException(string message, string title = "Conflict") : base(messa
 
     public ModuleOverlappingException(DateTime courseStart, DateTime courseEnd) 
-        : base($"Module must be within course dates: {courseStart} - {courseEnd}.")
+        : base($"Modulen måste ligga inom kursens datum: {courseStart} - {courseEnd}.")
```
Kanske borde vara couseStart.Date och courseEnd.Date, för att slippa klockslagen?

JorgenHagglund commented on Domain.Models/Exceptions/ConflictException.cs:
```diff
@@ -11,31 +11,31 @@ public ConflictException(string message, string title = "Conflict") : base(messa
 
 public class ActivityOverlapException(string range)
-    : ConflictException($"The time range: {range} overlaps with an other activity in this module.")
+    : ConflictException($"Tidsintervallet: {range} överlappar med en annan aktivitet i detta moment.")
```
detta moment => denna modul?

JorgenHagglund commented on Domain.Models/Exceptions/ConflictException.cs:
```diff
@@ -11,31 +11,31 @@ public ConflictException(string message, string title = "Conflict") : base(messa
 {
-    public DuplicateStudentInCourseException(string studentId, int courseId) : base($"Student {studentId} is already enrolled in course {courseId}.")
+    public DuplicateStudentInCourseException(string studentId, int courseId) : base($"Elev {studentId} är redan registrerad på kursen {courseId}.")
     {
```
Eleven?

JorgenHagglund commented on Domain.Models/Exceptions/ConflictException.cs:
```diff
@@ -11,31 +11,31 @@ public ConflictException(string message, string title = "Conflict") : base(messa
 public class DuplicateTeacherInCourseException : ConflictException
 {
-    public DuplicateTeacherInCourseException(string teacherId, int courseId) : base($"Teacher {teacherId} is already enrolled in course {courseId}.")
+    public DuplicateTeacherInCourseException(string teacherId, int courseId) : base($"Lärare {teacherId} är redan registrerad på kursen {courseId}.")
```
Läraren?

JorgenHagglund reviewed (approved):
Förövrigt ser det bra ut!

PhilippeLeopoldie committed:
Fix Swedish grammar in exception messages

Updated constructor messages in several exception classes for grammatical accuracy and clarity in Swedish. Adjusted date formatting in `ModuleOverlappingException` to ensure only the date part is included.

PhilippeLeopoldie commented:
> Förövrigt ser det bra ut!

Nu är allt fixat, tack för hjälpen! :)

JorgenHagglund reviewed (approved):
Perfekt!

ChornaIryna reviewed (approved):
Bra! Det finns även textsträngar i `ServiceBase ` som behöver översättas – vill du passa på att ta med dem också? 
<img width="1314" height="486" alt="image" src="https://github.com/user-attachments/assets/93448d43-364d-4c4e-b3f3-35db224708f5" />

PhilippeLeopoldie committed:
Translate error message in ValidateDateRange method

Updated the error message in the `ValidateDateRange` method from English to Swedish, changing "Start date must be before end date." to "Startdatum måste vara före slutdatum."

PhilippeLeopoldie commented:
> Bra! Det finns även textsträngar i `ServiceBase ` som behöver översättas – vill du passa på att ta med dem också? <img alt="image" width="1314" height="486" src="https://private-user-images.githubusercontent.com/148462332/486799426-93448d43-364d-4c4e-b3f3-35db224708f5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTczNDkxNjcsIm5iZiI6MTc1NzM0ODg2NywicGF0aCI6Ii8xNDg0NjIzMzIvNDg2Nzk5NDI2LTkzNDQ4ZDQzLTM2NGQtNGM0ZS1iM2YzLTM1ZGIyMjQ3MDhmNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwOFQxNjI3NDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kOTgwODg0YmM0N2QwMTE0ZTM0NTdjMjE5NzBmZWIxNDU1ZWNjMDNhMjdkNDA1ZDU5ODJiOWQyNDRkYWY1NTAwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ohWhhKgdNKn6WFVtIkSc-2gRF9giOsK9Rfodj9UC368">

Tack! Nu är det gjort :)

ahlgren1234 reviewed (approved):
Nice!
